### PR TITLE
[EBPF-356] Use io.Writer in ebpf map dumps 

### PR DESCRIPTION
**Pull request by gjulianm**
### What does this PR do?

This PR updates ebpf-manager to use io.Writer in Manager.DumpMaps, modifying all upstream callers properly.

### Motivation

Improve memory behavior so that maps do not need to be serialized into memory when being printed/transmitted.

### Additional Notes

Added a TestLogWriter type to wrap testing.T.Log so that it can be used as a io.Writer object.

### Describe how to test/QA your changes

All calls should be checked by the compiler, unit tests should be passing.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by hmahmood — APPROVED** [review]

File: cmd/system-probe/modules/network_tracer.go
Comment (leeavital) [inline cmd/system-probe/modules/network_tracer.go:207]: httpWriter itself implements writer, so you can pass it straight into DebugEbpfMaps. This will allow you to cut out the allocation in `var ebpfMaps bytes.Buffer`

File: pkg/ebpf/ebpftest/testlogwriter.go
Comment (leeavital) [inline pkg/ebpf/ebpftest/testlogwriter.go:23]: Did you find this useful? I could see this getting really noisy

**Review by leeavital — COMMENTED** [review]
Nice first (?) change!

File: cmd/system-probe/modules/network_tracer.go
Comment (brycekahle) [inline cmd/system-probe/modules/network_tracer.go:207]: I don't think there is a strong need to keep this a JSON string (`test` gets output as `"test"`). So I think you should be able to pass `w` directly to `nt.tracer.DebugEBPFMaps` and get rid of the `bytes.Buffer` altogether.

File: pkg/network/protocols/grpc/monitor_test.go
Comment (brycekahle) [inline pkg/network/protocols/grpc/monitor_test.go:516]: nit: move the `&` up here and use `tlw` below

File: pkg/network/tracer/connection/dump.go
Comment (brycekahle) [inline pkg/network/tracer/connection/dump.go:37]: should this file use `spew.Fdump` like the other files?

File: pkg/network/tracer/tracer.go
Comment (brycekahle) [inline pkg/network/tracer/tracer.go:719]: nit: error block spacing

File: pkg/network/tracer/tracer.go
Comment (brycekahle) [inline pkg/network/tracer/tracer.go:728]: nit: error block spacing

File: cmd/system-probe/modules/network_tracer.go
Comment (leeavital) [reply cmd/system-probe/modules/network_tracer.go:207]: I haven't used these endpoints myself in a while, but I could see it being handy to keep json output so we can pipe into `jq` to get the stats we care about. 

Though this change won't actually preserve that; I'm guessing it will write the contents of `ebpfMaps`  as one giant escaped JSON string.

Comment (brycekahle) [reply cmd/system-probe/modules/network_tracer.go:207]: Nothing below within the string is JSON formatted (it uses spew, see [example output](https://github.com/davecgh/go-spew?tab=readme-ov-file#sample-dump-output))

File: pkg/ebpf/ebpftest/testlogwriter.go
Comment (gjulianm) [reply pkg/ebpf/ebpftest/testlogwriter.go:23]: It's mostly to avoid having the pattern

```
var buffer bytes.Buffer
err := monitor.DumpMaps(&buffer, "name")

if err != nil {
  t.Logf("failed dumpMaps")
} else {
  t.Logf(buffer.String())
}
```

repeated all over the code. It's not a radical change, and from what I've seen in the output it doesn't seem to change things too much, but I can revert it if you prefer.

File: pkg/network/tracer/connection/dump.go
Comment (gjulianm) [reply pkg/network/tracer/connection/dump.go:37]: Yep, I missed that, changing it!

File: cmd/system-probe/modules/network_tracer.go
Comment (gjulianm) [reply cmd/system-probe/modules/network_tracer.go:207]: As Bryce says it just outputs the string as a quoted text, I kept it with JSON though just in case anyone was depending on that. If you think it's not important to have those quotes I'll change it to use the writer directly.

File: cmd/system-probe/modules/network_tracer.go
Comment (guyarb) [inline cmd/system-probe/modules/network_tracer.go:196]: who is responsible to write the output as json?

File: go.mod
Comment (guyarb) [inline go.mod:92]: we should merge this PR only after merging the PRs in ebpf-manager and releasing a new version

File: pkg/network/tracer/tracer.go
Comment (guyarb) [inline pkg/network/tracer/tracer.go:724]:
```suggestion
		io.WriteString(w, "usm_monitor:\n")
```

File: pkg/network/tracer/tracer_unsupported.go
Comment (guyarb) [inline pkg/network/tracer/tracer_unsupported.go:59]:
```suggestion
func (t *Tracer) DebugEBPFMaps(_ io.Writer, _ ...string) error {
	return ebpf.ErrNotImplemented
}
```

File: pkg/network/tracer/tracer_windows.go
Comment (guyarb) [inline pkg/network/tracer/tracer_windows.go:243]: The following fixes the revive linter warning, so you can remove the nolint line above
```suggestion
func (t *Tracer) DebugEBPFMaps(_ io.Writer, _ ...string) error {
```

**Review by guyarb — CHANGES_REQUESTED** [review]
In the description you mentioned
> Improve memory behavior so that maps do not need to be serialized into memory when being printed/transmitted.


Do we have any measurements? How can we ensure there is no reserved outcome?

File: cmd/system-probe/modules/network_tracer.go
Comment (gjulianm) [reply cmd/system-probe/modules/network_tracer.go:196]: In a previous discussion (https://github.com/DataDog/datadog-agent/pull/21598#discussion_r1430458762) Bryce said that it isn't needed to output as JSON, as in this case it's just serializing a string, the output will be the string but with quotes, that's why I removed the JSON call.

**Comment by gjulianm** [comment]
> In the description you mentioned
> 
> > Improve memory behavior so that maps do not need to be serialized into memory when being printed/transmitted.
> 
> Do we have any measurements? How can we ensure there is no reserved outcome?

I haven't done any measurements from before/after. I think the idea (@brycekahle correct me if I'm wrong) is to allow the dump methods to write directly to the output, so that it isn't needed to build the full string representation of the map in memory if the caller doesn't require it (http/stdout/stderr outputs) and instead it can be written in small chunks.

File: go.mod
Comment (gjulianm) [reply go.mod]: Merge on ebpf-manager was pending, I just bumped the version now to 0.4.0.

File: pkg/network/tracer/tracer_windows.go
Comment (brycekahle) [inline pkg/network/tracer/tracer_windows.go:242]: You can remove the `//nolint:revive` comment line since you fixed it

**Review by brycekahle — APPROVED** [review]

**Comment by guyarb** [comment]
> > allow the dump methods to write directly to the output, so that it isn't needed to build the full string representation of the map in memory if the caller doesn't require it (http/stdout/stderr outputs) and instead it can be written in small chunks.
> 
> Correct. The previous behavior is the worst case, so it can only get better.

That's true in theory, but I don't think we should rely on that. Mistakes can happen, and theories can fail
I think that's something we can easily measure and test its impact

**Review by derekwbrown — APPROVED** [review]

**Comment by gjulianm** [comment]
I ran some benchmarks in the ebpf-manager library, as that makes it simpler to see the effect of just this change without interference from the eBPF maps themselves (which are going to be there no matter the writing implementation).

The testing code for the previous version (v0.3.9) looks like this:

```go
m.DumpHandler = func(_ *Manager, mapName string, currentMap *ebpf.Map) string {
	var builder strings.Builder
	for j := 0; j < numDumps; j++ {
		builder.WriteString(dumpContents)
	}
	return builder.String()
}

for i := 0; i < b.N; i++ {
	_, err = m.DumpMaps("map_one")
	if err != nil {
		b.Fatal(err)
	}
}
```

For the new version with `io.Writer`, it looks like this: 

```go
m.DumpHandler = func(w io.Writer, _ *Manager, mapName string, currentMap *ebpf.Map) {
	for j := 0; j < numDumps; j++ {
		_, _ = io.WriteString(w, dumpContents)
	}
}

for i := 0; i < b.N; i++ {
	var buffer bytes.Buffer 
	err = m.DumpMaps(&buffer, "map_one")
	if err != nil {
		b.Fatal(err)
	}
}
```

This code runs for different values of numDump, so that we can simulate writing maps of different sizes. Here are the results for v0.3.9, with the previous implementation that used `strings.Builder` to build and return the string in the `DumpHandler` function:

Writes | ns/dump | Bytes allocated/dump | Allocs/dump
-- | -- | -- | --
1 | 92.65 | 16 | 2
3 | 136.0 | 80 | 4
10 | 204.1 | 328 | 6
100 | 757.5 | 2616 | 9
1000 | 7809 | 42488 | 16
10000 | 68111 | 359162 | 23

And here are the results for the new version, with `io.Writer` and using a `bytes.Buffer` as the `io.Writer`:

Writes | ns/dump | Bytes allocated/dump | Allocs/dump
-- | -- | -- | --
1 | 107.6 | 112 | 2
3 | 124.1 | 112 | 2
10 | 216.1 | 240 | 3
100 | 1127 | 2032 | 6
1000 | 8850 | 16368 | 9
10000 | 88400 | 262128 | 13


Note that while it's a little bit slower (although the difference is in the order of microseconds at most) the memory behavior is better, specially for larger maps. 

I did an extra test, using `io.Discard` as the `io.Writer`, to remove the influence of the backing writer on the results: 

Writes | ns/dump | Bytes allocated/dump | Allocs/dump
-- | -- | -- | --
1 | 63.40 | 0 | 0
3 | 77.09 | 0 | 0
10 | 117.5 | 0 | 0
100 | 596.2 | 0 | 0
1000 | 5572 | 0 | 0
10000 | 54041 | 0 | 0

What I interpret from this is that, by providing an `io.Writer` we remove the need to use the `strings.Builder` function and we just use as much memory as the underlying writer needs. For example, in the case of the HttpResponseWriter, we would write directly to its internal buffers, instead of creating our own and then copying them over.

**Comment by guyarb** [comment]
> I ran some benchmarks in the ebpf-manager library, as that makes it simpler to see the effect of just this change without interference from the eBPF maps themselves (which are going to be there no matter the writing implementation).
> 
> The testing code for the previous version (v0.3.9) looks like this:
> 
> ```go
> m.DumpHandler = func(_ *Manager, mapName string, currentMap *ebpf.Map) string {
> 	var builder strings.Builder
> 	for j := 0; j < numDumps; j++ {
> 		builder.WriteString(dumpContents)
> 	}
> 	return builder.String()
> }
> 
> for i := 0; i < b.N; i++ {
> 	_, err = m.DumpMaps("map_one")
> 	if err != nil {
> 		b.Fatal(err)
> 	}
> }
> ```
> 
> For the new version with `io.Writer`, it looks like this:
> 
> ```go
> m.DumpHandler = func(w io.Writer, _ *Manager, mapName string, currentMap *ebpf.Map) {
> 	for j := 0; j < numDumps; j++ {
> 		_, _ = io.WriteString(w, dumpContents)
> 	}
> }
> 
> for i := 0; i < b.N; i++ {
> 	var buffer bytes.Buffer 
> 	err = m.DumpMaps(&buffer, "map_one")
> 	if err != nil {
> 		b.Fatal(err)
> 	}
> }
> ```
> 
> This code runs for different values of numDump, so that we can simulate writing maps of different sizes. Here are the results for v0.3.9, with the previous implementation that used `strings.Builder` to build and return the string in the `DumpHandler` function:
> 
> Writes	ns/dump	Bytes allocated/dump	Allocs/dump
> 1	92.65	16	2
> 3	136.0	80	4
> 10	204.1	328	6
> 100	757.5	2616	9
> 1000	7809	42488	16
> 10000	68111	359162	23
> And here are the results for the new version, with `io.Writer` and using a `bytes.Buffer` as the `io.Writer`:
> 
> Writes	ns/dump	Bytes allocated/dump	Allocs/dump
> 1	107.6	112	2
> 3	124.1	112	2
> 10	216.1	240	3
> 100	1127	2032	6
> 1000	8850	16368	9
> 10000	88400	262128	13
> Note that while it's a little bit slower (although the difference is in the order of microseconds at most) the memory behavior is better, specially for larger maps.
> 
> I did an extra test, using `io.Discard` as the `io.Writer`, to remove the influence of the backing writer on the results:
> 
> Writes	ns/dump	Bytes allocated/dump	Allocs/dump
> 1	63.40	0	0
> 3	77.09	0	0
> 10	117.5	0	0
> 100	596.2	0	0
> 1000	5572	0	0
> 10000	54041	0	0
> What I interpret from this is that, by providing an `io.Writer` we remove the need to use the `strings.Builder` function and we just use as much memory as the underlying writer needs. For example, in the case of the HttpResponseWriter, we would write directly to its internal buffers, instead of creating our own and then copying them over.

Perfect! nice job

**Review by guyarb — APPROVED** [review]

**Pull request merged**
